tdrs-backend/tdpservice/parsers/util.py

```diff
@@ -281,3 +283,202 @@ def parse_and_validate(self, line, generate_error):
+            return models
+        elif model_name not in models.keys():
+            return None  # intentionally trigger the error_msg for unknown record type
+        else:
+            return models.get(model_name, models)
+
+
+'''
+text -> section YES
+text -> models{} YES
+text -> model YES
+datafile -> model
+    ^ section -> program -> model
+datafile -> text
+model -> text YES
+section -> text
+
+text**: input string from the header/file
+'''
+
+def get_program_models(str_prog, str_section):
+    """Return the models dict for a given program and section."""
+    return get_schema_options(program=str_prog, section=str_section, query='models')
+
+def get_program_model(str_prog, str_section, str_model):
+    """Return singular model for a given program, section, and name."""
+    return get_schema_options(program=str_prog, section=str_section, query='models', model_name=str_model)
+
+def get_section_reference(str_prog, str_section):
+    """Return the named section reference for a given program and section."""
+    return get_schema_options(program=str_prog, section=str_section, query='section')
+
+def get_text_from_df(df):
+    """Return the short-hand text for program, section for a given datafile."""
+    return get_schema_options("", section=df.section, query='text')
+
+def get_prog_from_section(str_section):
+    """Return the program type for a given section."""
+    # e.g., 'SSP Closed Case Data'
+    if str_section.startswith('SSP'):
+        return 'SSP'
+    elif str_section.startswith('Tribal'):
+        return 'TAN'  # problematic, do we need to infer tribal entirely from tribe/fips code?
+    else:
+        return 'TAN'
+
+    # TODO: if given a datafile (section), we can reverse back to the program b/c the
+    # section string has "tribal/ssp" in it, then process of elimination we have tanf
+
+def get_schema(line, section, program_type):
+    """Return the appropriate schema for the line."""
+    line_type = line[0:2]
+    return get_schema_options(program_type, section, query='models', model_name=line_type)
+
+def transform_to_months(quarter):
+    """Return a list of months in a quarter."""
+    match quarter:
+        case "Q1":
+            return ["Jan", "Feb", "Mar"]
+        case "Q2":
+            return ["Apr", "May", "Jun"]
+        case "Q3":
+            return ["Jul", "Aug", "Sep"]
+        case "Q4":
+            return ["Oct", "Nov", "Dec"]
+        case _:
+            raise ValueError("Invalid quarter value.")
+
+def month_to_int(month):
+    """Return the integer value of a month."""
+    return datetime.strptime(month, '%b').month
+
+
+def case_aggregates_by_month(df):
+    """Return case aggregates by month."""
+    section = str(df.section)  # section -> text
+    program_type = get_prog_from_section(section)  # section -> program_type -> text
+
+    # from datafile quarter, generate short month names for each month in quarter ala 'Jan', 'Feb', 'Mar'
+    month_list = transform_to_months(df.quarter)
```

Comment:
Kudos to @jtimpe for bringing this up with his testing. The df.quarter will represent fiscal quarter and will typically be ahead of the data's calendar dates.

Directly from @ADPennington :

> If frontend fiscal year = 2023 and fiscal quarter = 1 (Oct - Dec) and section = 1, then Header calendar year/quarter = 20224 and rpt_month_year = 202210, 202211, and 202212.

Meaning, we'll need to account for this delay/offset when relying about `df.quarter` so this is comment requires changes here.